tests/beta/interfaces/test_tape_tf.py

```diff
@@ -0,0 +1,741 @@
+            assert np.all(qtape.get_parameters() == [a, p[2], p[0], -p[2], p[1] + p[2]])
+
+            res = qtape.execute(device=dev)
+
+        expected = tf.cos(a) * tf.cos(p[1]) * tf.sin(p[0]) + tf.sin(a) * (
+            tf.cos(p[2]) * tf.sin(p[1]) + tf.cos(p[0]) * tf.cos(p[1]) * tf.sin(p[2])
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        res = tape.jacobian(res, p)
+        expected = np.array(
+            [
+                tf.cos(p[1]) * (tf.cos(a) * tf.cos(p[0]) - tf.sin(a) * tf.sin(p[0]) * tf.sin(p[2])),
+                tf.cos(p[1]) * tf.cos(p[2]) * tf.sin(a)
+                - tf.sin(p[1])
+                * (tf.cos(a) * tf.sin(p[0]) + tf.cos(p[0]) * tf.sin(a) * tf.sin(p[2])),
+                tf.sin(a)
+                * (tf.cos(p[0]) * tf.cos(p[1]) * tf.cos(p[2]) - tf.sin(p[1]) * tf.sin(p[2])),
+            ]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+        spy.assert_not_called()
+
+    def test_probability_differentiation(self, tol):
+        """Tests correct output shape and evaluation for a tape
+        with prob and expval outputs"""
+
+        dev = qml.device("default.qubit.tf", wires=2)
+        x = tf.Variable(0.543, dtype=tf.float64)
+        y = tf.Variable(-0.654, dtype=tf.float64)
+
+        with tf.GradientTape() as tape:
+            with QuantumTape() as qtape:
+                qml.RX(x, wires=[0])
+                qml.RY(y, wires=[1])
+                qml.CNOT(wires=[0, 1])
+                probs(wires=[0])
+                probs(wires=[1])
+
+            res = qtape.execute(dev)
+
+        expected = np.array(
+            [
+                [tf.cos(x / 2) ** 2, tf.sin(x / 2) ** 2],
+                [(1 + tf.cos(x) * tf.cos(y)) / 2, (1 - tf.cos(x) * tf.cos(y)) / 2],
+            ]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        res = tape.jacobian(res, [x, y])
+        expected = np.array(
+            [
+                [
+                    [-tf.sin(x) / 2, tf.sin(x) / 2],
+                    [-tf.sin(x) * tf.cos(y) / 2, tf.cos(y) * tf.sin(x) / 2],
+                ],
+                [
+                    [0, 0],
+                    [-tf.cos(x) * tf.sin(y) / 2, tf.cos(x) * tf.sin(y) / 2],
+                ],
+            ]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_ragged_differentiation(self, monkeypatch, tol):
+        """Tests correct output shape and evaluation for a tape
+        with prob and expval outputs"""
+        dev = qml.device("default.qubit.tf", wires=2)
+        x = tf.Variable(0.543, dtype=tf.float64)
+        y = tf.Variable(-0.654, dtype=tf.float64)
+
+        def _asarray(args, dtype=tf.float64):
+            res = [tf.reshape(i, [-1]) for i in args]
+            res = tf.concat(res, axis=0)
+            return tf.cast(res, dtype=dtype)
+
+        # we need to patch the asarray method on the device

```

Comment:
Oh, this is actually testing something that doesn't work at all in core. If you create a core QNode with ragged output (such as `return qml.probs(0), qml.probs(1, 2)`), passthru devices will choke, since we never considered this edge case.

A simple solution is simply to 'flatten' the ragged array; in NumPy, this is done via `np.hstack(ragged_array).flatten()`. In TensorFlow, the above `_asarray()` function does the same thing.

I made the change here in the test, rather than modifying the passthru devices directly --- I will make a note to modify the passthru devices once the tape moves out of beta

Reply:
have added a comment